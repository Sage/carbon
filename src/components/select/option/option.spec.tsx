import React from "react";
import TestRenderer from "react-test-renderer";
import { shallow, mount } from "enzyme";
import Option, { OptionProps } from ".";
import SelectListContext from "../__internal__/select-list-context";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function renderOption(props: OptionProps, renderer: any = shallow) {
  return renderer(<Option {...props} />);
}

describe("Option", () => {
  it("renders properly", () => {
    const props = { value: "1", text: "foo", children: "bar" };
    expect(renderOption(props, TestRenderer.create)).toMatchSnapshot();
  });

  describe("when no children are provided", () => {
    it("renders with the text prop as children", () => {
      const props = { value: "1", text: "foo" };
      expect(renderOption(props, TestRenderer.create)).toMatchSnapshot();
    });
  });

  describe("when isHighlighted prop is set", () => {
    it("then it should have expected background", () => {
      const props = { value: "1", text: "foo", isHighlighted: true };
      expect(renderOption(props, mount)).toHaveStyleRule(
        "background-color",
        "var(--colorsUtilityMajor200)"
      );
    });
  });

  describe("when hidden prop is set", () => {
    it("then it should have display set to none", () => {
      const props = { value: "1", text: "foo", hidden: true };
      expect(renderOption(props, mount)).toHaveStyleRule("display", "none");
    });
  });

  describe("when the element is inside the multiselect", () => {
    it("then it should have expected background", () => {
      const props = { value: "1", text: "foo" };
      expect(renderOption(props, mount)).toHaveStyleRule(
        "background-color",
        "var(--colorsUtilityMajor100)",
        { modifier: ":hover" }
      );
    });
  });

  describe("when the multiselectValues list contains the element value", () => {
    it("then the aria-selected attribute should be set to true", () => {
      const wrapper = mount(
        <SelectListContext.Provider value={{ multiselectValues: ["1"] }}>
          <Option value="1" text="foo" />
        </SelectListContext.Provider>
      );
      expect(wrapper.getDOMNode().getAttribute("aria-selected")).toBe("true");
    });
  });

  describe("when the multiselectValues list is present but does not contain the element value", () => {
    it("then the aria-selected attribute should be set to false", () => {
      const wrapper = mount(
        <SelectListContext.Provider value={{ multiselectValues: ["1", "2"] }}>
          <Option value="3" text="foo" />
        </SelectListContext.Provider>
      );
      expect(wrapper.getDOMNode().getAttribute("aria-selected")).toBe("false");
    });
  });

  describe("when the element is clicked", () => {
    it("then onSelect should be called with text and value", () => {
      const onSelect = jest.fn();
      const props = { value: "1", text: "foo", onSelect };
      const wrapper = renderOption(props, mount);
      wrapper.simulate("click");

      expect(onSelect).toHaveBeenCalledWith({
        text: props.text,
        value: props.value,
      });
    });
  });

  describe("tags on component", () => {
    it("includes correct component, element and role data tags", () => {
      const wrapper = mount(
        <Option
          value="1"
          text="one"
          data-component="foo"
          data-element="bar"
          data-role="baz"
        />
      );

      expect(wrapper.getDOMNode().getAttribute("data-component")).toEqual(
        "foo"
      );
      expect(wrapper.getDOMNode().getAttribute("data-element")).toEqual("bar");
      expect(wrapper.getDOMNode().getAttribute("data-role")).toEqual("baz");
    });
  });
});
