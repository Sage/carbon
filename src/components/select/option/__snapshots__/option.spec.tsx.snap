// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option renders properly 1`] = `
.c0 {
  cursor: pointer;
  box-sizing: border-box;
  line-height: 16px;
  padding: 12px 16px;
  width: 100%;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  background-color: var(--colorsUtilityMajor200);
}

.c0:hover {
  background-color: var(--colorsUtilityMajor100);
}

<li
  aria-selected={true}
  className="c0"
  data-component="option"
  onClick={[Function]}
  role="option"
>
  bar
</li>
`;

exports[`Option when no children are provided renders with the text prop as children 1`] = `
.c0 {
  cursor: pointer;
  box-sizing: border-box;
  line-height: 16px;
  padding: 12px 16px;
  width: 100%;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  background-color: var(--colorsUtilityMajor200);
}

.c0:hover {
  background-color: var(--colorsUtilityMajor100);
}

<li
  aria-selected={true}
  className="c0"
  data-component="option"
  onClick={[Function]}
  role="option"
>
  foo
</li>
`;
